File: src/org/nextras/orm/intellij/utils/OrmUtils.java
package org.nextras.orm.intellij.utils;

import com.intellij.psi.PsiElement;
import com.jetbrains.php.PhpIndex;
import com.jetbrains.php.lang.documentation.phpdoc.psi.PhpDocProperty;
import com.jetbrains.php.lang.psi.elements.*;
import com.jetbrains.php.lang.psi.resolve.types.PhpType;

import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;
import java.util.stream.Collectors;


public class OrmUtils
{
	public enum OrmClass
	{
		COLLECTION("\\Nextras\\Orm\\Collection\\ICollection"),
		MAPPER("\\Nextras\\Orm\\Mapper\\IMapper"),
		REPOSITORY("\\Nextras\\Orm\\Repository\\IRepository"),
		ENTITY("\\Nextras\\Orm\\Entity\\IEntity"),
		HAS_MANY("\\Nextras\\Orm\\Relationships\\HasMany");

		private String name;

		OrmClass(String name)
		{
			this.name = name;
		}

		public boolean is(PhpClass cls, PhpIndex index)
		{
			Collection<PhpClass> classes = index.getAnyByFQN(name);
			PhpClass instanceOf = classes.isEmpty() ? null : classes.iterator().next();
			if (instanceOf == null) {
				return false;
			}
			return instanceOf.getType().isConvertibleFrom(cls.getType(), index);
		}

	}


	public static Collection<PhpClass> findRepositoryEntities(Collection<PhpClass> repositories)
	{
		final Collection<PhpClass> entities = new HashSet<>(1);
		for (PhpClass repositoryClass : repositories) {

			Method entityNamesMethod = repositoryClass.findMethodByName("getEntityClassNames");
			if (entityNamesMethod == null) {
				return Collections.emptyList();
			}
			if (!(entityNamesMethod.getLastChild() instanceof GroupStatement)) {
				return Collections.emptyList();
			}
			if (!(((GroupStatement) entityNamesMethod.getLastChild()).getFirstPsiChild() instanceof PhpReturn)) {
				return Collections.emptyList();
			}
			if (!(((GroupStatement) entityNamesMethod.getLastChild()).getFirstPsiChild().getFirstPsiChild() instanceof ArrayCreationExpression)) {
				return Collections.emptyList();
			}
			ArrayCreationExpression arr = (ArrayCreationExpression) ((GroupStatement) entityNamesMethod.getLastChild()).getFirstPsiChild().getFirstPsiChild();
			final PhpIndex phpIndex = PhpIndex.getInstance(repositoryClass.getProject());
			for (PsiElement el : arr.getChildren()) {
				if (!(el.getFirstChild() instanceof ClassConstantReference)) {
					continue;
				}
				ClassConstantReference ref = (ClassConstantReference) el.getFirstChild();
				if (!ref.getName().equals("class")) {
					continue;
				}
				entities.addAll(PhpIndexUtils.getByType(ref.getClassReference().getType(), phpIndex));
			}
		}
		return entities;
	}


	public static Collection<PhpClass> findQueriedEntities(MemberReference ref)
	{
		PhpExpression classReference = ref.getClassReference();
		if (classReference == null) {
			return Collections.emptyList();
		}
		Collection<PhpClass> entities = new HashSet<>();
		PhpIndex phpIndex = PhpIndex.getInstance(ref.getProject());
		Collection<PhpClass> classes = PhpIndexUtils.getByType(classReference.getType(), phpIndex);
		Collection<PhpClass> repositories = classes.stream()
			.filter(cls -> OrmClass.REPOSITORY.is(cls, phpIndex))
			.collect(Collectors.toList());
		if (repositories.size() > 0) {
			entities.addAll(findRepositoryEntities(repositories));
		}
		entities.addAll(classes.stream().filter(cls -> OrmClass.ENTITY.is(cls, phpIndex)).collect(Collectors.toList()));
		return entities;
	}


	public static Collection<PhpClass> findQueriedEntities(MethodReference reference, String[] path)
	{
		if (path.length == 0) {
			return Collections.emptyList();
		}
		Collection<PhpClass> rootEntities;
		if (path.length == 1 || path[0].equals("this")) {
			rootEntities = findQueriedEntities(reference);
		} else {
			PhpIndex index = PhpIndex.getInstance(reference.getProject());
			rootEntities = PhpIndexUtils.getByType(new PhpType().add(path[0]), index);

		}
		if (rootEntities.size() == 0) {
			return Collections.emptyList();
		}
		if (path.length <= 1) {
			return rootEntities;
		}
		return findTargetEntities(rootEntities, path, 1);
	}


	private static Collection<PhpClass> findTargetEntities(Collection<PhpClass> currentEntities, String[] path, int pos)
	{
		if (path.length == (pos + 1)) {
			return currentEntities;
		}
		Collection<PhpClass> entities = new HashSet<>();
		for (PhpClass cls : currentEntities) {
			Field field = cls.findFieldByName(path[pos], false);
			if (!(field instanceof PhpDocProperty)) {
				continue;
			}
			addEntitiesFromField(entities, (PhpDocProperty) field);
		}
		return findTargetEntities(entities, path, pos + 1);
	}


	public static void addEntitiesFromField(Collection<PhpClass> entities, PhpDocProperty field)
	{
		PhpIndex index = PhpIndex.getInstance(field.getProject());
		for (String type : field.getType().getTypes()) {
			if (type.contains("Nextras\\Orm\\Relationship")) {
				continue;
			}
			if (type.endsWith("[]")) {
				type = type.substring(0, type.length() - 2);
			}
			for (PhpClass entityCls : PhpIndexUtils.getByType(new PhpType().add(type), index)) {
				if (!OrmClass.ENTITY.is(entityCls, index)) {
					continue;
				}
				entities.add(entityCls);
			}
		}
	}

}

File: src/org/nextras/orm/intellij/utils/OrmUtils.kt
package org.nextras.orm.intellij.utils

import com.jetbrains.php.PhpIndex
import com.jetbrains.php.lang.documentation.phpdoc.psi.PhpDocProperty
import com.jetbrains.php.lang.psi.elements.*
import com.jetbrains.php.lang.psi.resolve.types.PhpType
import java.util.*

object OrmUtils {
	enum class OrmClass constructor(private val className: String) {
		COLLECTION("\\Nextras\\Orm\\Collection\\ICollection"),
		MAPPER("\\Nextras\\Orm\\Mapper\\IMapper"),
		REPOSITORY("\\Nextras\\Orm\\Repository\\IRepository"),
		ENTITY("\\Nextras\\Orm\\Entity\\IEntity"),
		HAS_MANY("\\Nextras\\Orm\\Relationships\\HasMany");

		fun `is`(cls: PhpClass, index: PhpIndex): Boolean {
			val classes = index.getAnyByFQN(className)
			val instanceOf = when {
				classes.isEmpty() -> null
				else -> classes.iterator().next()
			} ?: return false
			return instanceOf.type.isConvertibleFrom(cls.type, index)
		}
	}

	fun findRepositoryEntities(repositories: Collection<PhpClass>): Collection<PhpClass> {
		val entities = HashSet<PhpClass>(1)
		for (repositoryClass in repositories) {

			val entityNamesMethod = repositoryClass.findMethodByName("getEntityClassNames") ?: return emptyList()
			if (entityNamesMethod.lastChild !is GroupStatement) {
				return emptyList()
			}
			if ((entityNamesMethod.lastChild as GroupStatement).firstPsiChild !is PhpReturn) {
				return emptyList()
			}
			if ((entityNamesMethod.lastChild as GroupStatement).firstPsiChild!!.firstPsiChild !is ArrayCreationExpression) {
				return emptyList()
			}
			val arr = (entityNamesMethod.lastChild as GroupStatement).firstPsiChild!!.firstPsiChild as ArrayCreationExpression?
			val phpIndex = PhpIndex.getInstance(repositoryClass.project)
			for (el in arr!!.children) {
				if (el.firstChild !is ClassConstantReference) {
					continue
				}
				val ref = el.firstChild as ClassConstantReference
				if (ref.name != "class") {
					continue
				}
				entities.addAll(PhpIndexUtils.getByType(ref.classReference!!.type, phpIndex))
			}
		}
		return entities
	}

	fun findQueriedEntities(ref: MemberReference): Collection<PhpClass> {
		val classReference = ref.classReference ?: return emptyList()
		val entities = HashSet<PhpClass>()
		val phpIndex = PhpIndex.getInstance(ref.project)
		val classes = PhpIndexUtils.getByType(classReference.type, phpIndex)
		val repositories = classes.filter { cls -> OrmClass.REPOSITORY.`is`(cls, phpIndex) }
		if (repositories.size > 0) {
			entities.addAll(findRepositoryEntities(repositories))
		}
		for (type in classReference.type.types) {
			if (!type.endsWith("[]")) {
				continue
			}
			val typeWithoutArray = PhpType().add(type.substring(0, type.length - 2))
			val maybeEntities = PhpIndexUtils.getByType(typeWithoutArray, phpIndex)
			entities.addAll(maybeEntities.filter { cls -> OrmClass.ENTITY.`is`(cls, phpIndex) })
		}

		return entities
	}

	fun findQueriedEntities(reference: MethodReference, path: Array<String>): Collection<PhpClass> {
		if (path.size == 0) {
			return emptyList()
		}
		val rootEntities: Collection<PhpClass>
		if (path.size == 1 || path[0] == "this") {
			rootEntities = findQueriedEntities(reference)
		} else {
			val index = PhpIndex.getInstance(reference.project)
			rootEntities = PhpIndexUtils.getByType(PhpType().add(path[0]), index)

		}
		if (rootEntities.size == 0) {
			return emptyList()
		}
		return if (path.size <= 1) {
			rootEntities
		} else findTargetEntities(rootEntities, path, 1)
	}

	private fun findTargetEntities(currentEntities: Collection<PhpClass>, path: Array<String>, pos: Int): Collection<PhpClass> {
		if (path.size == pos + 1) {
			return currentEntities
		}
		val entities = HashSet<PhpClass>()
		for (cls in currentEntities) {
			val field = cls.findFieldByName(path[pos], false) as? PhpDocProperty ?: continue
			addEntitiesFromField(entities, field)
		}
		return findTargetEntities(entities, path, pos + 1)
	}

	fun addEntitiesFromField(entities: MutableCollection<PhpClass>, field: PhpDocProperty) {
		val index = PhpIndex.getInstance(field.project)
		for (type in field.type.types) {
			if (type.contains("Nextras\\Orm\\Relationship")) {
				continue
			}
			val addType = if (type.endsWith("[]")) {
				type.substring(0, type.length - 2)
			} else {
				type
			}
			for (entityCls in PhpIndexUtils.getByType(PhpType().add(addType), index)) {
				if (!OrmClass.ENTITY.`is`(entityCls, index)) {
					continue
				}
				entities.add(entityCls)
			}
		}
	}
}
